stdlib/LinearAlgebra/src/symmetric.jl

```diff
@@ -453,25 +453,28 @@ function triu(A::Symmetric, k::Integer=0)
     end
 end
 
-for (T, trans, real) in [(:Symmetric, :transpose, :identity), (:Hermitian, :adjoint, :real)]
+for (T, trans, real) in [
+    (:Symmetric, :transpose, :identity),
+    (:(Hermitian{<:Union{Real,Complex}}), :adjoint, :real),
```

Comment:
It's slightly unfortunate that this specialization prevents this from being applicable to e.g. `Hermitian{<:AbstractMatrix{<:Union{Real,Complex}}`. Although I'm not sure how we might engineer a definition that's robust through multiple levels of nested `AbstractMatrix` so it would probably be hard to do something about this right now.

I'll also remark that this specialization should also be suitable for `dot(::Symmetric{<:Real},::Hermitian{<:Complex}}`, although currently it appears we miss this case. Would `LinearAlgebra.RealHermSymComplexHerm` be suitable here?

Reply:
I think this is a design flaw of LinearAlgebra. `Symmetric` should have been defined as real symmetric, and should be made a subtype of `Hermitian`, as mathematically a real symmetric matrix is just a particular case of a complex Hermitian matrix, and we'd get for free all of these particular cases without having to write a method for each specifically. Then one could also have a `GenericSymmetric` to deal with the more exotic cases if anyone needed them.

FWIW I did think about mixing real symmetric and complex Hermitian, but I decided against adding additional methods that were going to be rarely used. Of course, I know my opinion counts for very little here.